sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -90,6 +90,10 @@ trait PredicateHelper {
    * Returns true iff `expr` could be evaluated as a condition within join.
    */
   protected def canEvaluateWithinJoin(expr: Expression): Boolean = expr match {
+    case l: ListQuery =>
+      // Now pulling up IN predicates is deferred to `RewritePredicateSubquery`. As a result,
+      // `ListQuery`'s children can be empty before that rule and falls into the next case.
+      false
```

Comment:
Suggested text in the comment:

[NOT] IN (<ListQuery>) expression cannot be evaluated as part of a Join operator. Currently the only way to evaluate a [NOT] IN subquery is to convert it to a LeftSemi or LeftAnti by `RewritePredicateSubquery` rule.